internal/batches/ui/task_exec_tui.go

```diff
@@ -0,0 +1,391 @@
+			escapedLine := strings.ReplaceAll(lines[0], "%", "%%")
+			if len(lines) > 1 {
+				statusText = fmt.Sprintf("%s ...", escapedLine)
+			} else {
+				statusText = escapedLine
+			}
+		} else {
+			statusText = "..."
+		}
+	}
+
+	return statusText
+}
+
+type clock func() time.Time
+
+var defaultClock = time.Now
+
+func newTaskExecTUI(out *output.Output, verbose bool, numParallelism int) *taskExecTUI {
+	return &taskExecTUI{
+		out:            out,
+		verbose:        verbose,
+		numParallelism: numParallelism,
+
+		clock: defaultClock,
+
+		statuses:   map[*executor.Task]*taskStatus{},
+		statusBars: map[int]*taskStatus{},
+	}
+}
+
+type taskExecTUI struct {
+	// Used in tests only
+	forceNoSpinner bool
+
+	out *output.Output
+
+	verbose bool
+
+	progress      output.ProgressWithStatusBars
+	numStatusBars int
+
+	maxRepoName    int
+	numParallelism int
+
+	mu    sync.Mutex
+	clock clock
+
+	statuses   map[*executor.Task]*taskStatus
+	statusBars map[int]*taskStatus
+
+	finished int
+	errored  int
+}
+
+var _ executor.TaskExecutionUI = &taskExecTUI{}
+
+func (ui *taskExecTUI) Start(tasks []*executor.Task) {
+	for _, t := range tasks {
+		status := &taskStatus{}
+		if t.Path != "" {
+			status.displayName = t.Repository.Name + ":" + t.Path
+		} else {
+			status.displayName = t.Repository.Name
+		}
+
+		if len(status.displayName) > ui.maxRepoName {
+			ui.maxRepoName = len(status.displayName)
+		}
+
+		ui.statuses[t] = status
+	}
+
+	ui.numStatusBars = ui.numParallelism
+	if len(tasks) < ui.numStatusBars {
+		ui.numStatusBars = len(tasks)
+	}
+
+	statusBars := make([]*output.StatusBar, 0, ui.numStatusBars)
+	for i := 0; i < ui.numStatusBars; i++ {
+		statusBars = append(statusBars, output.NewStatusBar())
+	}
+
+	progressBars := []output.ProgressBar{
+		{
+			Label: fmt.Sprintf("Executing... (0/%d, 0 errored)", len(tasks)),
+			Max:   float64(len(tasks)),
+		},
+	}
+
+	opts := output.DefaultProgressTTYOpts.WithNoSpinner(ui.forceNoSpinner)
+	ui.progress = ui.out.ProgressWithStatusBars(progressBars, statusBars, opts)
+}
+
+func (ui *taskExecTUI) Success() {
+	ui.progress.Complete()
+}
+
+func (p *taskExecTUI) useFreeStatusBar(ts *taskStatus) (bar int, found bool) {
+	for i := 0; i < p.numStatusBars; i++ {
+		if _, ok := p.statusBars[i]; !ok {
+			p.statusBars[i] = ts
+			bar = i
+			found = true
+			return bar, found
+		}
+	}
+	return bar, found
+}
+
+func (p *taskExecTUI) findStatusBar(ts *taskStatus) (bar int, found bool) {
+	for i := 0; i < p.numStatusBars; i++ {
+		if status, ok := p.statusBars[i]; ok {
+			if ts == status {
+				bar = i
+				found = true
+				return bar, found
+			}
+		}
+	}
+
+	return bar, found
+}
+
+func (p *taskExecTUI) TaskStarted(task *executor.Task) {
+	p.mu.Lock()
+	defer p.mu.Unlock()
+
+	ts, ok := p.statuses[task]
+	if !ok {
+		panic("unknown task")
+	}
+
+	ts.startedAt = p.clock()
+
+	// Find free slot
+	bar, found := p.useFreeStatusBar(ts)
+	if !found {
+		panic("no available status bar found")
```

Comment:
Is a panic appropriate here (and below, and above, since I apparently clicked the wrong line)? We do return an error from this function: even though this is indicative of a logic error, could we continue as best we can with broken reporting?

Reply:
I just changed the lines to use `ui.out.Verbose` to print a warning so we don't crash.

The reason why I put the panic in is because the `p.statuses` access should never result in a miss. If it's a miss then someone called the methods in the wrong order and that should result in something blowing up.

For the `useFreeStatusBar` etc. there might be bugs lurking that could result in the panic. And the reason why I put the panic in is because these bugs are really hard to find and track down and a panic that dumps a stack trace might help.

But I did remove the `panic`s because I do agree with your point: we can continue as best as we can even if the UI is off. So my points would hold if this was essential functionality, but if someone is executing a batch spec for 5 hrs we should be okay if we don't show the right status.